homeassistant/components/idasen_desk/__init__.py

```diff
@@ -115,6 +116,24 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
     entry.async_on_unload(entry.add_update_listener(_async_update_listener))
 
+    @callback
+    def _async_bluetooth_callback(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a Bluetooth callback to ensure that a new BLEDevice is fetched."""
+        _LOGGER.debug("Bluetooth callback triggered. Reconnecting")
+        hass.async_create_task(coordinator.async_connect())
```

Comment:
Should this check if its already connected before creating a task?

Reply:
async_connect needs to be called when the BT callback is triggered, to ensure that it uses a new BLEDevice.

From https://developers.home-assistant.io/docs/bluetooth/ :

> Fetch a new BLEDevice from the bluetooth.async_ble_device_from_address API each time a connection is made. Alternatively, register a callback with bluetooth.async_register_callback and replace a cached BLEDevice each time a callback is received. The details of a BLEDevice object may change due to a change in the active adapter or environment.

Reply:
Oh. I need to remove that from the docs. The backend takes care of that now automatically

Reply:
So we can fetch the BLEDevice once from `bluetooth.async_ble_device_from_address` and it will stay valid?

Reply:
By the way, do you feel that it will cause any issue by calling connect every time the callback is called, even if already connected?
Because if not, I would prefer to merge this as it is for now, do the change you suggested separately and test it for a few days.

Since I had good results with this and we have a lot of users complaining about the desk getting unavailable, I would prefer to not risk not having this on the next release of HA.

Reply:
> So we can fetch the BLEDevice once from `bluetooth.async_ble_device_from_address` and it will stay valid?

When you call connect, a new one is fetched every time now, anyway, so you don't have to do anymore.

Reply:
> By the way, do you feel that it will cause any issue by calling connect every time the callback is called, even if already connected?
> 
> Because if not, I would prefer to merge this as it is for now, do the change you suggested separately and test it for a few days.
> 
> 
> 
> Since I had good results with this and we have a lot of users complaining about the desk getting unavailable, I would prefer to not risk not having this on the next release of HA.

If you are already connected, and then another proxy has a better signal, if you don't drop the connection and try to connect to the other one, it may fail in unexpected ways

Reply:
Problems with this kind of seems like the wrong solution, if the connection is dropping, you shouldn't have to wait for an advertisement to reconnect, as the address type should already be in the cache. When the disconnected call back is fired, you should probably set a flag and then restart the connection retry loop.

Reply:
I looked through the library a bit. If you refactor it to have a connection loop that runs while its expected to be connected and takes care of reconnect when its not, I think that will solve the issue.

You could refactor `async_connect` and `async_disconnect` to be the entry points to start and stop the reconnect loop and tear down the connection when its stopped.

Reply:
That already happened. The reconnection loop is already implemented:
When the connection drops, the integration calls async_connect() (see `async_set_updated_data` in the `IdasenDeskCoordinator`). 
On the library side, if `Desk.connect()` fails, it will retry after 30s, and keeps retrying until `Desk.disconnect` is called.

Still, it will keep failing to connect until I manually restarted the integration (which will in turn get a new BLEDevice). After the changes from this PR, it is always connecting automatically.

Reply:
But if you say that there is no need to fetch new BLEDevice, then I will try to dig deeper on why the automatic reconnection is not happening.

Reply:
In my test, I unplugged the ESPHome device hopping to see the reconnect loop kick in, and the desk reconnecting as soon as the ESPHome device connects again.
But it seems that the issue is that the bluetooth connect call just hangs. No error or timeout happens, so the reconnect loop never kicks in.

I'll try to debug this, but it may be an issue on ESPHome BT handling.

Reply:
I am not able to reproduce the issue in the `dev` branch. I guess that it was fixed on the recent updates to ESPHome bluetooth.